File: longest increasing subsequence.cpp
class Solution {
    public:
        int solve(int ind, int prev, vector<int>&nums, vector<vector<int>>&dp, int n){
            if(ind >= n)
                return 0;
            if(dp[ind][prev+1] != -1){
                return dp[ind][prev+1];
            }
            int pick=0, not_pick = 0;
            if(prev ==-1 || nums[ind] > nums[prev])
                pick = 1+ solve(ind+1, ind, nums, dp, n );
            not_pick = 0 + solve(ind+1, prev, nums, dp, n);
        
            return dp[ind][prev+1] = max(pick, not_pick);
        }
        int lengthOfLIS(vector<int>& nums) {
            int n = nums.size();
            vector<vector<int>>dp(n, vector<int>(n+1, -1));
            return solve(0, -1, nums, dp, n);
        }
    };

File: neetcode/300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    int solve(int []nums, int curr, int prev, int[][]dp, int n){

        //base case : curr > n -1
        if(curr > n-1) return 0;
        if(dp[curr][prev + 1] != -1) 
            return dp[curr][prev + 1];
        int notPick = 0 + solve(nums, curr+1, prev, dp, n);
        int pick = 0;
        if(prev == -1 || nums[curr] > nums[prev])
            pick = 1 + solve(nums, curr + 1, curr, dp, n);
        return dp[curr][prev + 1] = Math.max(notPick, pick);
    }
    public int lengthOfLIS(int[] nums) {
        
        int n = nums.length;
        int [][]dp = new int [n][n+1];
        //prev goes from -1 to n-1
        for(int i = 0; i < n; ++i) Arrays.fill(dp[i], -1);

        return solve(nums,0, -1,dp, n);
    }
}